packages/client/src/extension.mts

```diff
@@ -193,9 +195,47 @@ async function _activate(options: ActivateOptions): Promise<ExtensionApi> {
         client.onBlockFile(notifyUserOfBlockedFile),
     );
 
+    // If auto-correct is enabled, add the handler for it.
+    const config = vscode.workspace.getConfiguration();
+    if (config.get('cSpell.autoCorrect', false)) {
+        context.subscriptions.push(vscode.workspace.onDidChangeTextDocument(handleOnDidChangeTextDocument));
+    }
+
     performance.mark('registerCspellInlineCompletionProviders');
     await registerCspellInlineCompletionProviders(context.subscriptions).catch(() => undefined);
 
+    function handleOnDidChangeTextDocument({ document, contentChanges }: vscode.TextDocumentChangeEvent) {
+        // Check if the change should trigger auto-correct.
+        if (contentChanges.length === 0) return;
+        const lastChange = contentChanges[contentChanges.length - 1];
+        if (!AUTO_CORRECT_TRIGGER.test(lastChange.text)) return;
+
+        // Get cursor position prior to the change.
+        const position = lastChange.range.start;
+        const prevPosition = position.character > 0 ? new vscode.Position(position.line, position.character - 1) : undefined;
+        if (!prevPosition) return;
+
+        // Check if the cursor was at the end of a word.
+        const wordRange = document.getWordRangeAtPosition(prevPosition);
+        if (!wordRange) return;
+        if (!wordRange.end.isEqual(prevPosition.translate(0, 1))) return;
+
+        // Check if there were any cSpell diagnostics for that word.
+        const diagnostics = vscode.languages.getDiagnostics(document.uri);
```

Comment:
It is better to ask for the issues from `issueTracker.getSpellingIssues(document.uri)`.